OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs

```diff
@@ -22,22 +22,52 @@ public static bool IsCellBuildable(this World world, CPos cell, ActorInfo ai, Bu
 			if (!world.Map.Contains(cell))
 				return false;
 
-			var building = world.WorldActor.Trait<BuildingInfluence>().GetBuildingAt(cell);
-			if (building != null)
+			if (!bi.AllowInvalidPlacement)
 			{
-				if (ai == null)
-					return false;
+				// Replaceable actors are rare, so avoid initializing state unless we have to
+				var checkReplacements = ai != null && ai.HasTraitInfo<ReplacementInfo>();
+				HashSet<string> acceptedReplacements = null;
 
-				var replacementInfo = ai.TraitInfoOrDefault<ReplacementInfo>();
-				if (replacementInfo == null)
-					return false;
+				var foundActors = false;
+				foreach (var a in world.ActorMap.GetActorsAt(cell))
+				{
+					if (a == toIgnore)
+						continue;
+
+					// If this is potentially a replacement actor we must check *all* cell occupants
+					// before we know the placement is invalid
+					// Otherwise, we can bail immediately
+					if (!checkReplacements)
+						return false;
+
+					foundActors = true;
+					foreach (var r in a.TraitsImplementing<Replaceable>())
+					{
+						if (r.IsTraitDisabled)
+							continue;
+
+						if (acceptedReplacements == null)
+							acceptedReplacements = new HashSet<string>();
+
+						acceptedReplacements.UnionWith(r.Info.Types);
+					}
+				}
+
+				// The cell contained at least one (not ignored) actor
+				if (checkReplacements && foundActors)
```

Comment:
If `checkReplacements` is `false`, `foundActors` can never be true either, so this can be simplified to just `foundActors`. (But this is basically just a cosmetic change at this point, I don't really mind keeping it as is if you prefer that for readability.)